### PR TITLE
task/enhancements_and_bugs/color_selected_entities

### PR DESCRIPTION
**Pull request by BarAgam**
Selecting an entity will add a blue marker around it to indicate it's selected.

### CODE REVIEW

File: data/src/main/java/com/teamagam/gimelgimel/data/map/adapter/GeoEntityDataMapper.java [hunk 1]
```diff
@@ -37,24 +37,25 @@ public GeoEntity transform(String id, GeoContentData geoContentData) {
     }
 
     public ImageEntity transformIntoImageEntity(String id, PointGeometryData point) {
-        ImageEntity imageEntity = new ImageEntity(id, null, mGeometryMapper.transform(point));
+        ImageEntity imageEntity = new ImageEntity(id, null, mGeometryMapper.transform(point),
+                false);
```
Comment (teamagam5) [inline data/src/main/java/com/teamagam/gimelgimel/data/map/adapter/GeoEntityDataMapper.java:41]: A boolean parameter :scream: ...
I guess you don't like it either :smile:

File: data/src/main/java/com/teamagam/gimelgimel/data/map/repository/SelectedEntityDataRepository.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package com.teamagam.gimelgimel.data.map.repository;
+
+import com.teamagam.gimelgimel.domain.map.repository.SelectedEntityRepository;
+
+import javax.inject.Inject;
+
+public class SelectedEntityDataRepository implements SelectedEntityRepository {
+
+    private String mGeoEntity;
```
Comment (teamagam5) [inline data/src/main/java/com/teamagam/gimelgimel/data/map/repository/SelectedEntityDataRepository.java]: mGeoEntityId

File: domain/src/main/java/com/teamagam/gimelgimel/domain/map/SelectEntityInteractor.java [hunk 3]
```diff
@@ -3,36 +3,189 @@
...
 
     @Override
-    protected Observable buildUseCaseObservable() {
+    protected Iterable<SubscriptionRequest> buildSubscriptionRequests(
+            DataSubscriptionRequest.SubscriptionRequestFactory factory) {
+        DataSubscriptionRequest selectRelatedMessage =
+                factory.create(createSelectRelatedMessageObservable());
+
+        DataSubscriptionRequest colorSelection = factory.create(
+                Observable.just(mEntityId)
```
Comment (teamagam5) [inline domain/src/main/java/com/teamagam/gimelgimel/domain/map/SelectEntityInteractor.java]: Maybe extract a method?
Maybe even extract a method including the factory.create and then you get the form-
buildSubscriptionRequests()
{
  first = buildFirstRequest
  second = buildSecondRequest
  return as list(first, second)
}

File: domain/src/main/java/com/teamagam/gimelgimel/domain/map/SelectEntityInteractor.java [hunk 4]
```diff
@@ -3,36 +3,189 @@
...
+            removeOldSelection();
+            updateSelection(geoEntity);
+        }
+    }
+
+    private boolean isReselection(GeoEntity geoEntity) {
+        return geoEntity.getId().equals(mSelectedEntityRepository.getSelectedEntityId());
+    }
+
+    private void updateSelection(GeoEntity geoEntity) {
```
Comment (teamagam5) [inline domain/src/main/java/com/teamagam/gimelgimel/domain/map/SelectEntityInteractor.java]: removeOldSelection is called before this one

File: domain/src/main/java/com/teamagam/gimelgimel/domain/map/entities/symbols/BaseSymbol.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package com.teamagam.gimelgimel.domain.map.entities.symbols;
+
+
+public abstract class BaseSymbol implements Symbol {
```
Comment (teamagam5) [inline domain/src/main/java/com/teamagam/gimelgimel/domain/map/entities/symbols/BaseSymbol.java:4]: You can name it SelectableSymbol. I hate the word "base" :smiling_imp:

File: domain/src/main/java/com/teamagam/gimelgimel/domain/map/entities/symbols/Symbol.java [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 import com.teamagam.gimelgimel.domain.map.entities.interfaces.ISymbolVisitable;
 
-public interface Symbol extends ISymbolVisitable{
```
Comment (teamagam5) [inline domain/src/main/java/com/teamagam/gimelgimel/domain/map/entities/symbols/Symbol.java:5]: Was this a marker interface?

File: domain/src/main/java/com/teamagam/gimelgimel/domain/map/entities/symbols/Symbol.java [hunk 7]
```diff
@@ -2,6 +2,7 @@
...
+public interface Symbol extends ISymbolVisitable {
 
+    boolean isSelected();
```
Comment (teamagam5) [inline domain/src/main/java/com/teamagam/gimelgimel/domain/map/entities/symbols/Symbol.java:7]: I'm not sure that every symbol should have the "selectable" behavior. So maybe don't force it in the interface but as I suggested, convert the BaseSymbol class to SelectableSymbol class...

File: domain/src/main/java/com/teamagam/gimelgimel/domain/messages/poller/PolledMessagesProcessor.java [hunk 8]
```diff
@@ -148,6 +152,7 @@ private void mapEntityToMessage(BaseMessageGeo message) {
         }
 
         private void displayGeoEntity(GeoEntity geoEntity) {
+            mGeoEntitiesRepository.add(geoEntity);
```
Comment (teamagam5) [inline domain/src/main/java/com/teamagam/gimelgimel/domain/messages/poller/PolledMessagesProcessor.java:155]: Kinda duplicate don't u think? I mean, now every time I want to display an entity I must remember to first add it to the entities repo and then to show it using the displayed entities repo.
We have the same issue with messages I think...

File: data/src/main/java/com/teamagam/gimelgimel/data/map/adapter/GeoEntityDataMapper.java (hunk 1 above)
Comment (teamagam5) [reply data/src/main/java/com/teamagam/gimelgimel/data/map/adapter/GeoEntityDataMapper.java:41]: Now I finished the CR and I think maybe we can avoid it but I need your guidance. Let's talk about it :)

**Review by teamagam5 — APPROVED** [review]

File: domain/src/main/java/com/teamagam/gimelgimel/domain/messages/poller/PolledMessagesProcessor.java (hunk 8 above)
Comment (BarAgam) [reply domain/src/main/java/com/teamagam/gimelgimel/domain/messages/poller/PolledMessagesProcessor.java:155]: We've talked about it, you're right - we can refactor the code to allow drawing only through an interactor that will handle it by itself (and hide the DisplayRepository from classes outside its package).
That being said, we'll have to postpone it :<

**Pull request merged**